File: src/api/java/maestro/api/schematic/RotatedSchematic.java
package maestro.api.schematic;

import java.util.List;
import java.util.stream.Collectors;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.block.state.BlockState;

public class RotatedSchematic implements ISchematic {

    private final ISchematic schematic;
    private final Rotation rotation;
    private final Rotation inverseRotation;

    public RotatedSchematic(ISchematic schematic, Rotation rotation) {
        this.schematic = schematic;
        this.rotation = rotation;
        // I don't think a 14 line switch would improve readability
        this.inverseRotation = rotation.getRotated(rotation).getRotated(rotation);
    }

    @Override
    public boolean inSchematic(int x, int y, int z, BlockState currentState) {
        return schematic.inSchematic(
                rotateX(x, z, widthX(), lengthZ(), inverseRotation),
                y,
                rotateZ(x, z, widthX(), lengthZ(), inverseRotation),
                rotate(currentState, inverseRotation));
    }

    @Override
    public BlockState desiredState(
            int x, int y, int z, BlockState current, List<BlockState> approxPlaceable) {
        return rotate(
                schematic.desiredState(
                        rotateX(x, z, widthX(), lengthZ(), inverseRotation),
                        y,
                        rotateZ(x, z, widthX(), lengthZ(), inverseRotation),
                        rotate(current, inverseRotation),
                        rotate(approxPlaceable, inverseRotation)),
                rotation);
    }

    @Override
    public void reset() {
        schematic.reset();
    }

    @Override
    public int widthX() {
        return flipsCoordinates(rotation) ? schematic.lengthZ() : schematic.widthX();
    }

    @Override
    public int heightY() {
        return schematic.heightY();
    }

    @Override
    public int lengthZ() {
        return flipsCoordinates(rotation) ? schematic.widthX() : schematic.lengthZ();
    }

    /** Wether {@code rotation} swaps the x and z components */
    private static boolean flipsCoordinates(Rotation rotation) {
        return rotation == Rotation.CLOCKWISE_90 || rotation == Rotation.COUNTERCLOCKWISE_90;
    }

    /** The x component of x,y after applying the rotation */
    private static int rotateX(int x, int z, int sizeX, int sizeZ, Rotation rotation) {
        return switch (rotation) {
            case NONE -> x;
            case CLOCKWISE_90 -> sizeZ - z - 1;
            case CLOCKWISE_180 -> sizeX - x - 1;
            case COUNTERCLOCKWISE_90 -> z;
        };
    }

    /** The z component of x,y after applying the rotation */
    private static int rotateZ(int x, int z, int sizeX, int sizeZ, Rotation rotation) {
        return switch (rotation) {
            case NONE -> z;
            case CLOCKWISE_90 -> x;
            case CLOCKWISE_180 -> sizeZ - z - 1;
            case COUNTERCLOCKWISE_90 -> sizeX - x - 1;
        };
    }

    private static BlockState rotate(BlockState state, Rotation rotation) {
        if (state == null) {
            return null;
        }
        return state.rotate(rotation);
    }

    private static List<BlockState> rotate(List<BlockState> states, Rotation rotation) {
        if (states == null) {
            return null;
        }
        return states.stream().map(s -> rotate(s, rotation)).collect(Collectors.toList());
    }
}

File: common/src/main/maestro/schematic/RotatedSchematic.kt
package maestro.schematic

import net.minecraft.world.level.block.Rotation
import net.minecraft.world.level.block.state.BlockState

class RotatedSchematic(
    schematic: ISchematic,
    private val rotation: Rotation,
) : TransformedSchematic(schematic) {
    // I don't think a 14 line switch would improve readability
    private val inverseRotation: Rotation = rotation.getRotated(rotation).getRotated(rotation)

    override fun inSchematic(
        x: Int,
        y: Int,
        z: Int,
        currentState: BlockState?,
    ): Boolean =
        schematic.inSchematic(
            rotateX(x, z, widthX(), lengthZ(), inverseRotation),
            y,
            rotateZ(x, z, widthX(), lengthZ(), inverseRotation),
            rotate(currentState, inverseRotation),
        )

    override fun desiredState(
        x: Int,
        y: Int,
        z: Int,
        current: BlockState?,
        approxPlaceable: List<BlockState>?,
    ): BlockState =
        rotate(
            schematic.desiredState(
                rotateX(x, z, widthX(), lengthZ(), inverseRotation),
                y,
                rotateZ(x, z, widthX(), lengthZ(), inverseRotation),
                rotate(current, inverseRotation),
                rotate(approxPlaceable, inverseRotation),
            ),
            rotation,
        )!!

    override fun widthX(): Int = if (flipsCoordinates(rotation)) schematic.lengthZ() else schematic.widthX()

    override fun lengthZ(): Int = if (flipsCoordinates(rotation)) schematic.widthX() else schematic.lengthZ()

    companion object {
        /** Whether [rotation] swaps the x and z components */
        private fun flipsCoordinates(rotation: Rotation): Boolean =
            rotation == Rotation.CLOCKWISE_90 || rotation == Rotation.COUNTERCLOCKWISE_90

        /** The x component of x,z after applying the rotation */
        private fun rotateX(
            x: Int,
            z: Int,
            sizeX: Int,
            sizeZ: Int,
            rotation: Rotation,
        ): Int =
            when (rotation) {
                Rotation.NONE -> x
                Rotation.CLOCKWISE_90 -> sizeZ - z - 1
                Rotation.CLOCKWISE_180 -> sizeX - x - 1
                Rotation.COUNTERCLOCKWISE_90 -> z
            }

        /** The z component of x,z after applying the rotation */
        private fun rotateZ(
            x: Int,
            z: Int,
            sizeX: Int,
            sizeZ: Int,
            rotation: Rotation,
        ): Int =
            when (rotation) {
                Rotation.NONE -> z
                Rotation.CLOCKWISE_90 -> x
                Rotation.CLOCKWISE_180 -> sizeZ - z - 1
                Rotation.COUNTERCLOCKWISE_90 -> sizeX - x - 1
            }

        private fun rotate(
            state: BlockState?,
            rotation: Rotation,
        ): BlockState? = state?.rotate(rotation)

        private fun rotate(
            states: List<BlockState>?,
            rotation: Rotation,
        ): List<BlockState>? = states?.map { rotate(it, rotation)!! }
    }
}
